Added translations with flow support

pnicolli opened the pull request:
These translations come as separate js objects, for example:

```javascript
const it = {
  where: 'Dove:',
};
```

In the `config.json` file, the user may specify the preferred language as a 2-characters string (e.g. 'it', 'en', 'fr', ...).

As a bonus, I added flow support for translation dictionaries, in order to have help from the editor in adding translations and using translation. Flow will in fact check if we are typing the correct keys for our translation dictionaries.

Future improvements may include:

- Interpolation of translated strings (i.e. having variables inside strings, in order to have more complex translations)
- Unit tests for the translation function

pnicolli committed:
Added first translations. Installed flow.

pnicolli committed:
Fixed flow configuration, added remaining translations

pnicolli committed:
Updated config.sample.json

Takeno commented on utils/discord-sender.js:
```diff
@@ -1,16 +1,35 @@
+    )} *${event.applicationCloseRemaining}*`;
+
     if (process.env.DRY_RUN) {
+        logger.info(`DRY_RUN selected.
```
I think we should use `logger.debug` for DRY_RUN env.

pnicolli replied:
Yeah, that's fine, I'll change this.

Takeno commented on utils/discord-sender.js:
```diff
@@ -1,16 +1,35 @@
     logger.info('Posting event to discord: ', event);
+    const t = translate(config.language);
+    const description = formatDiscordDescription(event, config.language);
+    const content = `${t(
```
I think this is not so much readable.
I thought two ways:
* Template tag ``i18n`Key ${1} key ${2}` -> i18n([ 'Key ', ' key ', '' ], 1, 2 )``
* printf-like syntax

What do you think about?

pnicolli replied:
I understand it's not perfectly readable. Instead of a template syntax, I would use interpolation (which needs to be developed for this translation system), which would be used like this in this scenario:
```javascript
const content = t(
  'content',
  { /* object with variables */ },
  'default message',
);
```

Template strings are cool too, but that would require using weird strings as a key for language dictionaries, for example the string `'Applications will close in: *'`.
In addition to this, using a template string still forces you to keep the position of keys and variables inside the string for every language, which I am not sure is the case for all languages for this sentence.

pnicolli replied:
@Takeno I added a commit right now with the new interpolation function. Does it look better now? I feel like this is the best solution because it is now possible to have the placeholders in different spots for each language, in case it's needed.

Meanwhile, I also updated the README with informations on how to configure languages.

Takeno commented:
Basically, I would keep texts as readable as possible and really easy to be changed in future.
Yeah, some of them can be considered "future improvements" but I like to start in the right way :)

pnicolli committed:
Added interpolation function and updated README

pnicolli committed:
Rebase on new master. Added dry-run script.

pnicolli commented:
@Takeno Rebased on master after merging #6

pnicolli committed:
Updated README

pnicolli committed:
Removed defaultMessage, using en language as default

pnicolli commented:
Updated PR: the default message used is now automatically the english one. Also, refactored `discord-sender.js` in order to separate message creation logic from the rest of the code.

pnicolli committed:
Moved translations to json

Takeno commented on src/i18n/index.js:
```diff
@@ -0,0 +1,70 @@
+// @flow
+import en from '../translations/en.json';
+import it from '../translations/it.json';
```
Can we try reading files from directory? 
e.g.
```
const files = fs.readdirSync('../translations');
const translations = files.reduce((dict, file) => 
   const locale = file.split('.').shift(); // en | it | en_US
   dict[locale] = require(`../translations/${file}`;
   return dict;
), {});
```

Takeno replied:
In this way we can manage localization by adding or removing files from `translations` directory

Takeno commented on src/utils/translate.js:
```diff
@@ -0,0 +1,11 @@
+    key: $Keys<Language>,
+    substitutions?: {[string]: string},
+): string {
+    return translate(config.language)(key, substitutions);
```
Based on previous comment, I think is better check language before starting.

Maybe a `localeExists(language)` that throws an error. If app starts, we're safe with localized strings

Takeno commented:
A simple json validator for translations:
```
const fs = require('fs');
const i18nKeys = ['when', 'where', 'closing', 'content'];

const files = fs.readdirSync('../translations');

files.forEach(file =>  {
    const json = require(`../translations/${file}`);

    i18nKeys.forEach(k => {
        if(!json[k])
            throw new Error(`${file} doesn't contain '${k}' key`);
    })
});
```

pnicolli committed:
Added flow-typed dependencies

pnicolli commented:
I was wondering... if we are checking whether each translation has all the needed keys, otherwise we are stopping the execution, should we also require the presence of the whole translation file?
From what we have so far (including your latest suggestions), we get that:

- if a translation is missing, english is used instead (for example: `config.json` has the spanish language, but we have no `es.json` translation file)
- if a translation is incomplete, we stop execution and throw an error

I feel a little incoherence here. What do you think about this?

pnicolli committed:
Added strict runtime check to translation files

Takeno commented:
Yep, you're right.
My idea was to create a dev-command to check the translations files. Once the scripts, it should simply load configured language file or load `en.json`, if it doesn't exists.

My philosophy is to be able to create new translations and check them without run the entire server (probably we also can run the format-functions to let see the results).
In this way we can have non-technical collaborators (for this and other projects). 

What do you think? Is it just ugly overhead?

pnicolli commented:
I get your point, it makes sense. I have an idea, let me wire that up, expect a followup commit soon 😃

pnicolli committed:
Load only the needed language and check it at runtime. Created separate test script for language files.

pnicolli committed:
Load english language when configured language is not available.

pnicolli commented:
@Takeno Pushed the new commit. Now the bot loads only the needed language, or english if the configured language is not available, and checks it at runtime. I created a separate test script for all language files.
